Fixes #29092 - Set Active Tab for AuthSourceExternal form

aaditi-jain opened the pull request:
Bug:
https://projects.theforeman.org/issues/29092

Steps to reproduce bug:
1. log in to the dashboard as a administrator user   
1. On the dashboard, click on 'Administrator' in the left nav-bar and then click on 'Authentication sources'  
1. Then on the card titled as 'External', click on the icon at top-right  
1. Click on the 'edit' option from the dropdown
1. You will notice that on the page that loads (/auth_source_externals/{:id}/edit), there is no active tab and the tab contents are empty

Before Fix Screenshot:

![29092_before](https://user-images.githubusercontent.com/1745111/79999101-24e55380-84d9-11ea-82a5-9b1595586985.png)

After Fix Screenshot:

![29092 _after](https://user-images.githubusercontent.com/1745111/79999043-1434dd80-84d9-11ea-8b2c-9a4fcc398e2b.png)


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

aaditi-jain committed:
Fixes #29092 - Set Active Tab for AuthSourceExternal form

rahulbajaj0509 commented:
[test foreman]

rabajaj0509 commented:
[test foreman]

tbrisker commented:
OK to test

rabajaj0509 reviewed (dismissed):
I have tested the PR and it works as expected! LGTM, will still wait for final review from @sharvit  @tbrisker . Meanwhile you might want to rebase your PR :)

ezr-ondrej commented on app/views/auth_source_externals/_form.html.erb:
```diff
@@ -2,16 +2,21 @@
   <ul class="nav nav-tabs" data-tabs="tabs">
     <% if show_location_tab? %>
-      <li><a href="#locations" data-toggle="tab"><%= _("Locations") %></a></li>
+      <li class='<%=tab_classes_for_edit_auth_source_external[:location]%>'>
```
```suggestion
      <li class='active'>
```

ezr-ondrej commented on app/views/auth_source_externals/_form.html.erb:
```diff
@@ -2,16 +2,21 @@
     <% end %>
     <% if show_organization_tab? %>
-      <li><a href="#organizations" data-toggle="tab"><%= _("Organizations") %></a></li>
+      <li class='<%=tab_classes_for_edit_auth_source_external[:organization]%>'>
```
```suggestion
      <li class="<%= 'active' unless show_location_tab? %>">
```

ezr-ondrej commented on app/views/auth_source_externals/_form.html.erb:
```diff
@@ -2,16 +2,21 @@
-    <%= render 'taxonomies/loc_org_tabs', :f => f, :obj => @auth_source_external %>
-</div>
+    <%= render 'taxonomies/loc_org_tabs', :f => f, :obj => @auth_source_external,
+      html_options: { tab_classes: tab_classes_for_edit_auth_source_external } %>
```
```suggestion
      :active_tab => show_location_tab? ? :location : :organization %>
```

ezr-ondrej commented on app/views/taxonomies/_loc_org_tabs.html.erb:
```diff
@@ -1,10 +1,11 @@
 <% select_options ||= {} %>
 <% html_options ||= {} %>
+<% html_options[:tab_classes] ||= {} %>
```
```suggestion
<% active_tab ||= nil %>
```

ezr-ondrej commented on app/views/taxonomies/_loc_org_tabs.html.erb:
```diff
@@ -1,10 +1,11 @@
 
 <% if show_location_tab? %>
-  <div class="tab-pane" id="locations">
+  <div class='<%= "tab-pane #{html_options[:tab_classes][:location]}" %>' id="locations">
```
```suggestion
  <div class="tab-pane<%= ' active' if active_tab == :location %>" id="locations">
```

ezr-ondrej commented on app/views/taxonomies/_loc_org_tabs.html.erb:
```diff
@@ -17,7 +18,7 @@
 
 <% if show_organization_tab? %>
-  <div class="tab-pane" id="organizations">
+  <div class='<%= "tab-pane #{html_options[:tab_classes][:organization]}" %>' id="organizations">
```
```suggestion
  <div class="tab-pane<%= ' active' if active_tab == :organization %>" id="organizations">
```

ezr-ondrej reviewed (commented):
I don't think the helper is necessary, I've left idea how we could aproach it without the helper.
I'm not against the helper though :+1:
If you feel like you would like to have the helper, just please keep the formating
- <%=(space!) ... (space!)
- even though it's unfortunate, we use `=>` hash style
- (nitpick) html should use `"` (double quotes) e.g. `class="<%= .. %>"`

aaditi-jain commented:
@rahulbajaj0509 @ezr-ondrej : Thank you for the review :-)
@ezr-ondrej : The approach that you suggested in the review seemed more readable to me. However (imho) using the helper will help keep the logic extracted at one place and view files neater. Hence continued to use the helper.
I updated the hash syntax and added the missing spaces as suggested.

kgaikwad reviewed (changes requested):
It's good to have helper methods to extract the complex logic out of the view. With the current changes, I didn't find any good reason to add helper method so I agree with @ezr-ondrej. 

Few observations from code changes:
- Both .erb files still having these conditions in place and this method is returning just same class name with taxonomy mapping.
- IIUC, while rendering the form, it gives call to this `tab_classes_for_edit_auth_source_external` method thrice. Every time it is going to execute `show_location_tab?` method. That's why it is important to check what this method does inside?

Either by using any variable at ERB level or if you would like to keep helper then returning the output like
`{ :show_location => show_location_tab?, :show_organization => show_organization_tab?}` would be enough. You can use this value throughout form ERB.

@aaditi-jain, @ezr-ondrej, what do you think?

aaditi-jain commented:
@kgaikwad :
Thank you for looking into this :)
I am okay to put the entire logic in the view files (like @ezr-ondrej suggested), if putting it into helper seems like a over-kill.

Also about the code changes suggested in the above [comment](https://github.com/theforeman/foreman/pull/7582#pullrequestreview-398492905), here is an explanation: 

From the existing [auth_source_externals/_form.rb](https://github.com/theforeman/foreman/blob/10e7aa5c16283d90dd92d679051ac2853d742069/app/views/auth_source_externals/_form.html.erb#L4) code, it seems that there can be a condition where current_user has show_organization permission and **NOT** show_location permission. In that case, the user should see only an **active** organization tab. This is the condition that is handled in the helper method in this MR. We can also add conditions and tackle it in the view files as @ezr-ondrej suggested.

However, if I updated the helper method to return `{ :show_location => show_location_tab?, :show_organization => show_organization_tab? }` as suggested in the comment, it will require that we add logic in view files also to check the value of :show_location / :show_organization key. That is, the logic will spread both in the helper and in the view files, which kills the purpose of introducing the helper.

Hope that explains the changes. Let me know your thoughts

kgaikwad commented:
> However, if I updated the helper method to return `{ :show_location => show_location_tab?, :show_organization => show_organization_tab? }` as suggested in the comment, it will require that we add logic in view files also to check the value of :show_location / :show_organization key.

@aaditi-jain,
Yes, it will definitely need changes at view level to check the value by using either variable or helper_method. 
I will prefer a variable to avoid unnecessarily multiple calls to methods `show_location_tab? & show_organization_tab?` and rest of the conditional handling will remain as @ezr-ondrej  suggested.

aaditi-jain commented:
The more I look at the code, the more I do not see a good reason to move the logic to views from the helper
Here is my reasoning:
1. We are introducing 4 conditions across 2 view files if we move the logic to view. This is reduced to 1 condition statement contained at in a single method by using helper.

1. Considering the "complexity" of the helper method, which was pointed out [earlier](https://github.com/theforeman/foreman/pull/7582#pullrequestreview-398492905) as a reason to shift to views, I found that there are similar helper methods used generously across the code base. Hence, using helpers seems to be consistent with the code pattern followed in the codebase too. Here are examples from existing code base:

- TaxonomyHelper#edit_taxonomy_path
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/taxonomy_helper.rb#L52
- TaxonomyHelper#mismatches_taxonomies_path
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/taxonomy_helper.rb#L76
- TemplatesHelper#hide_resource_type_input
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/templates_helper.rb#L57
- EditorHelper#template_name_attribute
https://github.com/theforeman/foreman/blob/33a5b34b2a984c3689639f3df08f23e83b2104aa/app/helpers/editor_helper.rb#L2

Taking cue from an earlier suggestion, I have introduced a variable in view file to store the output of the existing helper method and updated the code.
@kgaikwad, looking forward to know your thoughts :)

ezr-ondrej commented on app/views/taxonomies/_loc_org_tabs.html.erb:
```diff
@@ -1,10 +1,11 @@
 
 <% if show_location_tab? %>
-  <div class="tab-pane" id="locations">
+  <div class="<%= "tab-pane #{html_options[:tab_classes][:location]}" %>" id="locations">
```
This is equal and safes string interpolation, mainly it's bit more readable IMHO.
```suggestion
  <div class="tab-pane <%= html_options[:tab_classes][:location] %>" id="locations">
```

ezr-ondrej commented on app/views/taxonomies/_loc_org_tabs.html.erb:
```diff
@@ -17,7 +18,7 @@
 
 <% if show_organization_tab? %>
-  <div class="tab-pane" id="organizations">
+  <div class="<%= "tab-pane #{html_options[:tab_classes][:organization]}" %>" id="organizations">
```
`^^`

ezr-ondrej reviewed (commented):
I don't have strong opinion on the helper :)
It works for me eighter way :+1:

kgaikwad commented on app/views/auth_source_externals/_form.html.erb:
```diff
@@ -1,17 +1,23 @@
+    <% tab_classes = tab_classes_for_edit_auth_source_external %>
     <% if show_location_tab? %>
-      <li><a href="#locations" data-toggle="tab"><%= _("Locations") %></a></li>
+      <li class="<%= tab_classes[:location] %>">
```
External auth source form is having only two tabs (i.e locations & organizations). It seems like locations tab is primary one. 
Can't we just add class `active` directly as to show this tab we already had this condition check?

```suggestion
      <li class="active">
```

ezr-ondrej replied:
~I don't believe this is possible. This template is used in other places and those have more tabs :)~
EDIT: I was too quick I though I'm commenting on different file, this is a good idea :+1:

kgaikwad reviewed (changes requested):
Thank you @aaditi-jain for providing the pointer files and quick updates.
I still vote for changes without a helper method :-) but I am OK with the current changes.

Just added one inline comment.  
To make it consistent with other changes, you can ignore below comment but thought to share it.

kgaikwad commented on test/helpers/auth_source_external_helper_test.rb:
```diff
@@ -0,0 +1,37 @@
+    context 'user with neither view_locations nor view_organization access' do
+      it 'returns a empty hash' do
+        setup_user 'none'
+        assert_equal ({}), tab_classes_for_edit_auth_source_external
```
```suggestion
        assert_empty tab_classes_for_edit_auth_source_external
```

kgaikwad commented on test/helpers/auth_source_external_helper_test.rb:
```diff
@@ -0,0 +1,37 @@
+    context 'user with view_organizations permission only' do
+      it 'returns class for organization tab' do
+        setup_user 'view', 'organizations'
+        assert_equal ({ :organization => 'active' }), tab_classes_for_edit_auth_source_external
```
```suggestion
        assert_not_empty tab_classes_for_edit_auth_source_external[:organization]
```

aaditi-jain replied:
Similar to as https://github.com/theforeman/foreman/pull/7582#discussion_r419672478

kgaikwad commented on test/helpers/auth_source_external_helper_test.rb:
```diff
@@ -0,0 +1,37 @@
+    context 'user with view_locations permission only' do
+      it 'returns class for location tab' do
+        setup_user 'view', 'locations'
+        assert_equal ({ :location => 'active' }), tab_classes_for_edit_auth_source_external
```
```suggestion
       assert_not_empty tab_classes_for_edit_auth_source_external[:location]
```

aaditi-jain replied:
@kgaikwad The suggested changes will not cover the method's functionality completely.
We are not only testing here that location tab has 'active' classes, but also that organization tab does not have any 'active' class. 
`assert_not_empty tab_classes_for_edit_auth_source_external[:location]` only checks that :location tab is active. It does not check that the :organization key is not present.

kgaikwad replied:
> It does not check that the :organization key is not present.

Suggested this change because for any change in tab_classes, one will need to touch these test cases unnecessarily. What do you think about adding multiple assert statements in one test? Here, for example -

```
assert_equal [:location], tab_classes_for_edit_auth_source_external.keys
assert_not_empty tab_classes_for_edit_auth_source_external[:location]
```

kgaikwad commented on test/helpers/auth_source_external_helper_test.rb:
```diff
@@ -0,0 +1,37 @@
+    context 'user with view_locations and view_organizations permission' do
+      it 'returns class for location tab' do
+        as_admin do
+          assert_equal ({ :location => 'active' }), tab_classes_for_edit_auth_source_external
```
```suggestion
         assert_not_empty tab_classes_for_edit_auth_source_external[:location]
```

aaditi-jain replied:
Same as https://github.com/theforeman/foreman/pull/7582#discussion_r419672478

kgaikwad reviewed (changes requested):
Thank you @aaditi-jain!
It would be good to verify using specific key value than comparing the whole object in tests so added small nitpicks. Apart of this everything looks good for first version :+1:

aaditi-jain commented:
> It would be good to verify using specific key value than comparing the whole object in tests so added small nitpicks. Apart of this everything looks good for first version +1

@kgaikwad I have added my opinion about the suggested changes. Looking forward to know your thoughts on it :). Also not really sure what do we mean by `first version` mentioned above.

kgaikwad reviewed (approved):
It would be good to avoid hash comparison but it works. 
I don't want to block this PR further due to those changes. 

Thank you @aaditi-jain for quick updates on review comments :ok_hand: :tada:

@rahulbajaj0509, @ezr-ondrej  - any additional comments from your side?

rabajaj0509 reviewed (approved):
I think this one is a good to go! thanks @aaditi-jain

ezr-ondrej reviewed (approved):
I'm ok with this. As this is bugfix, it should get it in before branching.
Right @tbrisker ?

tbrisker reviewed (approved):
Thanks @aaditi-jain and all reviewers!

tbrisker commented:
2.0 cherry-pick - d98547067

aaditi-jain commented:
Thank you @kgaikwad, @ezr-ondrej for the reviews :smiley:. Thank you @rahulbajaj0509 too for helping in setting up foreman on my local and in picking up a first issue.

alexjfisher commented:
This PR seems related to an issue I've discovered after updating to 2.0.1 (from 1.24).  If I'm logged into a specific org/location (such as 'default location'), then I get a 404 when I click on my ldap authentication source.  I've opened a new ticket. 
 See https://projects.theforeman.org/issues/30319
